qutebrowser/config/configtypes.py

```diff
@@ -1151,6 +1162,17 @@ class UserStyleSheet(File):
     def __init__(self):
         super().__init__(none_ok=True)
 
+    def transform(self, value):
+        if not value:
+            return None
+        path = super().transform(value)
+        path = os.path.expandvars(path)
```

Comment:
Those tests currently use Python's [unittest.mock](https://docs.python.org/3/library/unittest.mock.html) library, and `os_path` is the mock object which replaces the real `os.path.<function>` calls in the `conftypes.py` module.

One rather surprising thing about mocks is that any call to them return *another* mock object if you didn't configure them to do something else via `the_mock.return_value = ...` or `the_mock.side_effect = ...`.

At the beginning of the test, we do `os_path.expanduser.side_effect = lambda x: x.replace('~', '/home/foo')`. This assigns an anonymous function (a [lambda](https://docs.python.org/3/reference/expressions.html#lambda)) to `expanduser` which returns the given string with `~` replaced by `/home/foo`.

However, we don't do the same with `os_path.expandvars` but call `expandvars` from our code - which causes `os_path.expandvars(...)` to return another mock object (a `MagicMock`), which fails our test:

```
>       assert self.t.transform('~/foobar') == '/home/foo/foobar'
E       assert <MagicMock name='path.expandvars()' id='139799983029048'> == '/home/foo/foobar'
```

The solution here is to provide some stub for `expandvars`. I got it working with this diff:

```diff
diff --git a/qutebrowser/config/configtypes.py b/qutebrowser/config/configtypes.py
index 4b0b5b7..2ecb303 100644
--- a/qutebrowser/config/configtypes.py
+++ b/qutebrowser/config/configtypes.py
@@ -803,6 +803,7 @@ class File(BaseType):
         if not value:
             return None
         value = os.path.expanduser(value)
+        value = os.path.expandvars(value)
         if not os.path.isabs(value):
             if standarddir.config():
                 abspath = os.path.join(standarddir.config(), value)
@@ -1166,7 +1167,6 @@ class UserStyleSheet(File):
         if not value:
             return None
         path = super().transform(value)
-        path = os.path.expandvars(path)
         if os.path.isabs(path):
             return QUrl.fromLocalFile(path)
         else:
diff --git a/tests/config/test_configtypes.py b/tests/config/test_configtypes.py
index c515c6c..0f7b355 100644
--- a/tests/config/test_configtypes.py
+++ b/tests/config/test_configtypes.py
@@ -1398,6 +1398,7 @@ class TestFile:
     def test_transform(self, os_path):
         """Test transform."""
         os_path.expanduser.side_effect = lambda x: x.replace('~', '/home/foo')
+        os_path.expandvars.side_effect = lambda x: x
         assert self.t.transform('~/foobar') == '/home/foo/foobar'
         os_path.expanduser.assert_called_once_with('~/foobar')
```

But looking at the tests again now, maybe I should rewrite them to use real files in a temporary directory instead - I agree mocks can be hard to grasp.